src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -78,18 +78,19 @@ public function validate($entity, Constraint $constraint)
             }
         }
 
-        $class = $em->getClassMetadata(get_class($entity));
+        $entityClass = $constraint->entityClass ?: get_class($entity);
+        $class = $em->getClassMetadata($entityClass);
         /* @var $class \Doctrine\Common\Persistence\Mapping\ClassMetadata */
 
         $criteria = array();
         $hasNullValue = false;
 
-        foreach ($fields as $fieldName) {
-            if (!$class->hasField($fieldName) && !$class->hasAssociation($fieldName)) {
-                throw new ConstraintDefinitionException(sprintf('The field "%s" is not mapped by Doctrine, so it cannot be validated for uniqueness.', $fieldName));
+        foreach ($fields as $formField => $entityField) {
+            if (!$class->hasField($entityField) && !$class->hasAssociation($entityField)) {
+                throw new ConstraintDefinitionException(sprintf('The field "%s" is not mapped by Doctrine, so it cannot be validated for uniqueness.', $entityField));
             }
 
-            $fieldValue = $class->reflFields[$fieldName]->getValue($entity);
+            $fieldValue = $class->reflFields[is_int($formField) ? $entityField : $formField]->getValue($entity);
```

Comment:
we need to consider https://github.com/symfony/symfony/issues/22592#issuecomment-298336733 here?

> we use PropertyAccess component instead of ReflectionProperty to read the field value from current object (maintain the metadata verification and the DTO fields must be readable too),

Reply:
Yes indeed

Reply:
But that would add a dependency, as we just read fields maybe reflection is simpler.

Reply:
why not :P